File: src/main/java/org/cubexmc/railway/listener/PlayerMoveListener.java
package org.cubexmc.railway.listener;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.event.vehicle.VehicleExitEvent;
import org.cubexmc.railway.Railway;
import org.cubexmc.railway.manager.StopManager;
import org.cubexmc.railway.model.Line;
import org.cubexmc.railway.model.Stop;
import org.cubexmc.railway.util.SchedulerUtil;
import org.cubexmc.railway.util.TextUtil;
import org.cubexmc.railway.util.AdventureUtil;

public class PlayerMoveListener implements Listener {

    private final Railway plugin;
    private final Map<UUID, Object> actionBarTasks = new HashMap<>();
    private final Map<UUID, Object> titleTasks = new HashMap<>();
    private final Map<UUID, String> playerStopMap = new HashMap<>();

    public PlayerMoveListener(Railway plugin) {
        this.plugin = plugin;
    }

    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
    public void onPlayerMove(PlayerMoveEvent event) {
        if (isSameBlock(event.getFrom(), event.getTo())) return;
        Player player = event.getPlayer();
        if (!player.hasPermission("railway.use")) return;
        Location target = event.getTo() != null ? event.getTo() : player.getLocation();
        processPlayerLocation(player, target);
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onVehicleExit(VehicleExitEvent event) {
        if (!(event.getExited() instanceof Player)) {
            return;
        }
        Player player = (Player) event.getExited();
        if (!player.hasPermission("railway.use")) {
            return;
        }
        SchedulerUtil.entityRun(plugin, player, () -> processPlayerLocation(player, player.getLocation()), 1L, -1L);
    }

    private void showEnterStopUI(Player player, Line line, Stop stop, StopDisplayContext context) {
        if (player.isInsideVehicle()) {
            return;
        }
        int eta = plugin.getLineServiceManager().estimateNextEtaSeconds(line.getId(), stop.getId());
        sendStopTitle(player, line, stop, context, eta);
    }

    private void startActionBarTask(Player player, Line line, Stop stop, StopDisplayContext context) {
        UUID pid = player.getUniqueId();
        if (player.isInsideVehicle()) {
            return;
        }
        Object task = SchedulerUtil.entityRun(plugin, player, () -> {
            if (!player.isOnline()) { cancelTasks(pid); return; }
            if (player.isInsideVehicle()) { cancelTasks(pid); return; }
            Location loc = player.getLocation();
            if (!stop.isInStop(loc)) { cancelTasks(pid); return; }
            int eta = plugin.getLineServiceManager().estimateNextEtaSeconds(line.getId(), stop.getId());
            String actionbar = applyPlaceholders(context.actionbarTemplate, line, stop, context.lastStop, context.nextStop, context.terminalStop, eta);
            AdventureUtil.sendActionBar(player, actionbar);
        }, 0L, 20L);
        actionBarTasks.put(pid, task);
    }

    private Line findFirstLineContaining(Stop stop) {
        if (stop == null) return null;
        for (Line line : plugin.getLineManager().getAllLines()) {
            if (line.containsStop(stop.getId())) {
                return line;
            }
        }
        return null;
    }

    private void startTitleTask(Player player, Line line, Stop stop, StopDisplayContext context) {
        UUID pid = player.getUniqueId();
        cancelTitleTask(pid);
        if (player.isInsideVehicle()) {
            return;
        }
        if (!context.alwaysShow) {
            return;
        }

        Object task = SchedulerUtil.entityRun(plugin, player, () -> {
            if (!player.isOnline()) { cancelTasks(pid); return; }
            if (player.isInsideVehicle()) { cancelTasks(pid); return; }
            Location loc = player.getLocation();
            if (!stop.isInStop(loc)) { cancelTasks(pid); return; }
            int eta = plugin.getLineServiceManager().estimateNextEtaSeconds(line.getId(), stop.getId());
            sendStopTitle(player, line, stop, context, eta);
        }, Math.max(1L, context.intervalTicks), Math.max(1L, context.intervalTicks));
        titleTasks.put(pid, task);
    }

    private void processPlayerLocation(Player player, Location location) {
        StopManager stopManager = plugin.getStopManager();
        Stop stop = stopManager.getStopContainingLocation(location);
        UUID pid = player.getUniqueId();
        String current = playerStopMap.get(pid);

        if (stop != null) {
            if (!stop.getId().equals(current)) {
                playerStopMap.put(pid, stop.getId());
                cancelTasks(pid);
                Line line = findFirstLineContaining(stop);
                if (line != null && !player.isInsideVehicle()) {
                    StopDisplayContext context = buildDisplayContext(line, stop);
                    showEnterStopUI(player, line, stop, context);
                    startActionBarTask(player, line, stop, context);
                    startTitleTask(player, line, stop, context);
                }
            } else if (!player.isInsideVehicle()) {
                boolean hasActionTask = actionBarTasks.containsKey(pid);
                boolean hasTitleTask = titleTasks.containsKey(pid);
                boolean needsTitleTask = plugin.isStopContinuousAlways() && !hasTitleTask;
                if (!hasActionTask || needsTitleTask) {
                    Line line = findFirstLineContaining(stop);
                    if (line != null) {
                        StopDisplayContext context = buildDisplayContext(line, stop);
                        showEnterStopUI(player, line, stop, context);
                        if (!hasActionTask) {
                            startActionBarTask(player, line, stop, context);
                        }
                        if (needsTitleTask) {
                            startTitleTask(player, line, stop, context);
                        }
                    }
                }
            } else {
                cancelTasks(pid);
                clearTitles(player);
            }
        } else if (current != null) {
            playerStopMap.remove(pid);
            cancelTasks(pid);
            clearTitles(player);
        }
    }

    private void clearTitles(Player player) {
        AdventureUtil.clearTitle(player);
        AdventureUtil.sendActionBar(player, "");
    }

    private StopDisplayContext buildDisplayContext(Line line, Stop stop) {
        StopManager stopManager = plugin.getStopManager();
        String stopId = stop.getId();
        String prevId = line.getPreviousStopId(stopId);
        String nextId = line.getNextStopId(stopId);

        Stop prev = prevId != null ? stopManager.getStop(prevId) : null;
        Stop next = nextId != null ? stopManager.getStop(nextId) : null;

        Stop terminal = null;
        List<String> stopIds = line.getOrderedStopIds();
        if (!stopIds.isEmpty()) {
            String terminalId = stopIds.get(stopIds.size() - 1);
            terminal = stopManager.getStop(terminalId);
        }

        String title;
        String subtitle;
        String actionbar;

        if (prev == null) {
            title = plugin.getStartStopTitle();
            subtitle = plugin.getStartStopSubtitle();
            actionbar = plugin.getStartStopActionbar();
        } else if (next == null) {
            title = plugin.getEndStopTitle();
            subtitle = plugin.getEndStopSubtitle();
            actionbar = plugin.getStopContinuousActionbar();
        } else {
            title = plugin.getStopContinuousTitle();
            subtitle = plugin.getStopContinuousSubtitle();
            actionbar = plugin.getStopContinuousActionbar();
        }

        Map<String, String> custom = stop.getCustomTitle("stop_continuous");
        if (custom != null) {
            if (custom.containsKey("title")) title = custom.get("title");
            if (custom.containsKey("subtitle")) subtitle = custom.get("subtitle");
            if (custom.containsKey("actionbar")) actionbar = custom.get("actionbar");
        }

        StopDisplayContext context = new StopDisplayContext();
        context.lastStop = prev;
        context.nextStop = next;
        context.terminalStop = terminal;
        context.titleTemplate = title;
        context.subtitleTemplate = subtitle;
        context.actionbarTemplate = actionbar;
        context.alwaysShow = plugin.isStopContinuousAlways();
        context.intervalTicks = Math.max(1, plugin.getStopContinuousInterval());
        if (context.alwaysShow) {
            context.fadeIn = 0;
            context.fadeOut = 0;
            int stay = plugin.getStopContinuousStay();
            context.stay = Math.max(stay, context.intervalTicks + 1);
        } else {
            context.fadeIn = plugin.getStopContinuousFadeIn();
            context.stay = plugin.getStopContinuousStay();
            context.fadeOut = plugin.getStopContinuousFadeOut();
        }
        return context;
    }

    private String applyPlaceholders(String template, Line line, Stop current, Stop last, Stop next, Stop terminal, int etaSeconds) {
        String result = TextUtil.replacePlaceholders(template, line, current, last, next, terminal, plugin.getLineManager());
        String etaString = etaSeconds >= 0 ? String.valueOf(etaSeconds) : "--";
        String etaFormatted = etaSeconds >= 0 ? formatEta(etaSeconds) : "--:--";
        result = result.replace("{eta_seconds}", etaString);
        result = result.replace("{eta_formatted}", etaFormatted);
        return result;
    }

    private void sendStopTitle(Player player, Line line, Stop stop, StopDisplayContext context, int etaSeconds) {
        String title = applyPlaceholders(context.titleTemplate, line, stop, context.lastStop, context.nextStop, context.terminalStop, etaSeconds);
        String subtitle = applyPlaceholders(context.subtitleTemplate, line, stop, context.lastStop, context.nextStop, context.terminalStop, etaSeconds);
        AdventureUtil.sendTitle(player, title, subtitle, context.fadeIn, context.stay, context.fadeOut);
    }

    private String formatEta(int etaSeconds) {
        int minutes = Math.max(0, etaSeconds) / 60;
        int seconds = Math.max(0, etaSeconds) % 60;
        return String.format("%d:%02d", minutes, seconds);
    }

    private void cancelTasks(UUID pid) {
        cancelActionBarTask(pid);
        cancelTitleTask(pid);
    }

    private void cancelActionBarTask(UUID pid) {
        Object task = actionBarTasks.remove(pid);
        if (task != null) {
            SchedulerUtil.cancelTask(task);
        }
    }

    private void cancelTitleTask(UUID pid) {
        Object task = titleTasks.remove(pid);
        if (task != null) {
            SchedulerUtil.cancelTask(task);
        }
    }

    private boolean isSameBlock(Location a, Location b) {
        if (a == null || b == null) return true;
        return a.getBlockX() == b.getBlockX() && a.getBlockY() == b.getBlockY() && a.getBlockZ() == b.getBlockZ();
    }

    private static class StopDisplayContext {
        Stop lastStop;
        Stop nextStop;
        Stop terminalStop;
        String titleTemplate;
        String subtitleTemplate;
        String actionbarTemplate;
        boolean alwaysShow;
        int fadeIn;
        int stay;
        int fadeOut;
        int intervalTicks;
    }
}

File: src/main/java/org/cubexmc/metro/listener/PlayerMoveListener.kt
package org.cubexmc.metro.listener

import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.BaseComponent
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.ChatColor
import org.bukkit.Location
import org.bukkit.entity.Minecart
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerMoveEvent
import org.bukkit.metadata.FixedMetadataValue
import org.bukkit.persistence.PersistentDataType
import org.cubexmc.metro.Metro
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.model.Stop
import org.cubexmc.metro.util.MetroConstants
import org.cubexmc.metro.util.SchedulerUtil
import org.cubexmc.metro.util.TextUtil
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.max

/**
 * 监听玩家移动事件，用于检测玩家进入停靠区
 */
class PlayerMoveListener(private val plugin: Metro) : Listener {

    /** 记录玩家当前所在的停靠区ID */
    private val playerInStopMap: MutableMap<UUID, String> = ConcurrentHashMap()

    /** 记录持续显示信息的任务ID */
    private val continuousInfoTasks: MutableMap<UUID, Any> = ConcurrentHashMap()

    /** 记录专门的ActionBar显示任务ID */
    private val actionBarTasks: MutableMap<UUID, Any> = ConcurrentHashMap()

    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
    fun onPlayerMove(event: PlayerMoveEvent) {
        // 只有玩家从一个方块移动到另一个方块时才检查，优化性能
        if (isSameBlock(event.from, event.to)) {
            return
        }

        val player = event.player
        if (!player.hasPermission("railway.use")) {
            return
        }

        // 检查玩家是否在矿车内，如果在矿车内则不显示站台信息
        if (isInMetroMinecart(player)) {
            val playerId = player.uniqueId
            if (playerInStopMap.remove(playerId) != null) {
                cancelContinuousInfoTask(playerId)
                cancelActionBarTask(playerId)
            }
            return
        }

        // 玩家不在矿车内，正常处理站台信息
        val stop = plugin.stopManager.getStopContainingLocation(player.location)

        val playerId = player.uniqueId
        val currentStopId = playerInStopMap[playerId]

        // 检查玩家是否进入了新的停靠区
        if (stop != null) {
            val stopId = stop.id
            if (stopId != currentStopId) {
                // 玩家进入了新的停靠区
                playerInStopMap[playerId] = stopId

                // 取消原来的持续显示任务
                cancelContinuousInfoTask(playerId)

                // 启动新的持续显示任务
                if (plugin.configFacade.isStopContinuousTitleEnabled()) {
                    startContinuousInfoTask(player, stop)
                }
            }
        } else if (currentStopId != null) {
            // 玩家离开了停靠区
            playerInStopMap.remove(playerId)
            cancelContinuousInfoTask(playerId)
            cancelActionBarTask(playerId) // 取消ActionBar任务

            // 立即清除title和actionbar显示
            player.resetTitle()
            // 发送空的ActionBar来清除显示
            player.spigot().sendMessage(ChatMessageType.ACTION_BAR, *TextComponent.fromLegacyText(""))

            // 清除首次运行标记
            player.removeMetadata("metro_first_run_$currentStopId", plugin)
        }
    }

    /**
     * 检查两个位置是否在同一方块内
     */
    private fun isSameBlock(from: Location?, to: Location?): Boolean {
        if (from == null || to == null) {
            return true
        }

        return from.blockX == to.blockX && from.blockY == to.blockY && from.blockZ == to.blockZ
    }

    /**
     * 启动持续显示停靠区信息的任务
     */
    private fun startContinuousInfoTask(player: Player, stop: Stop) {
        val playerId = player.uniqueId

        // 取消已有的ActionBar任务
        cancelActionBarTask(playerId)

        val lineManager = plugin.lineManager
        val boardableLines = plugin.lineSelectionService.getBoardableLines(stop)

        val line: Line
        if (boardableLines.isEmpty()) {
            val terminalLines = plugin.lineSelectionService.getTerminalLines(stop)
            if (terminalLines.isEmpty()) {
                return
            }
            line = terminalLines[0]
        } else if (boardableLines.size > 1) {
            startMultiLineInfoTask(player, stop, boardableLines)
            return
        } else {
            line = boardableLines[0]
        }

        val config = plugin.configFacade
        val timings = TitleTimings.from(config)

        // 准备信息内容（在任务外提前准备，以便ActionBar任务可以使用）
        // 获取前一站和下一站信息
        val stopManager = plugin.stopManager
        val lastStopId = line.getPreviousStopId(stop.id)
        val nextStopId = line.getNextStopId(stop.id)
        val lastStop = if (lastStopId == null) null else stopManager.getStop(lastStopId)
        val nextStop = if (nextStopId == null) null else stopManager.getStop(nextStopId)

        // 获取终点站信息
        val stopIds = line.orderedStopIds
        val terminalStop = if (stopIds.isEmpty()) null else stopManager.getStop(stopIds[stopIds.size - 1])

        // 确定站点类型并获取对应配置
        val isStartStop = lastStop == null // 没有上一站，是始发站
        val isEndStop = nextStop == null // 没有下一站，是终点站

        var title = config.getStopContinuousTitle(isStartStop, isEndStop)
        var subtitle = config.getStopContinuousSubtitle(isStartStop, isEndStop)
        var actionbar = config.getStopContinuousActionbar(isStartStop, isEndStop)

        val customTitle = stop.getCustomTitle("stop_continuous")
        if (customTitle != null) {
            customTitle["title"]?.let { title = it }
            customTitle["subtitle"]?.let { subtitle = it }
            customTitle["actionbar"]?.let { actionbar = it }
        }

        val translatedTitle = {
            translateStopTemplate(title, line, stop, lastStop, nextStop, terminalStop, lineManager)
        }
        val translatedSubtitle = {
            translateStopTemplate(subtitle, line, stop, lastStop, nextStop, terminalStop, lineManager)
        }
        val actionbarComponent = {
            TextComponent.fromLegacyText(
                translateStopTemplate(actionbar, line, stop, lastStop, nextStop, terminalStop, lineManager),
            )
        }

        showStopInfo(player, stop, timings, translatedTitle, translatedSubtitle, actionbarComponent)
    }

    private fun startMultiLineInfoTask(player: Player, stop: Stop, boardableLines: List<Line>) {
        val config = plugin.configFacade
        val timings = TitleTimings.from(config)

        val translatedTitle = {
            ChatColor.translateAlternateColorCodes(
                '&',
                plugin.languageManager.getMessage("interact.multi_line_title", mapOf("stop_name" to stop.name)),
            )
        }
        val translatedSubtitle = {
            ChatColor.translateAlternateColorCodes(
                '&',
                plugin.languageManager.getMessage(
                    "interact.multi_line_subtitle",
                    mapOf("count" to boardableLines.size.toString()),
                ),
            )
        }
        val actionbarComponent = {
            TextComponent.fromLegacyText(
                ChatColor.translateAlternateColorCodes(
                    '&',
                    plugin.languageManager.getMessage(
                        "interact.multi_line_actionbar",
                        mapOf("routes" to buildBoardableRouteSummary(stop, boardableLines)),
                    ),
                ),
            )
        }

        showStopInfo(player, stop, timings, translatedTitle, translatedSubtitle, actionbarComponent)
    }

    private fun showStopInfo(
        player: Player,
        stop: Stop,
        timings: TitleTimings,
        translatedTitle: () -> String,
        translatedSubtitle: () -> String,
        actionbarComponent: () -> Array<BaseComponent>,
    ) {
        if (timings.alwaysShow) {
            startAlwaysShowTasks(player, stop, timings, translatedTitle, translatedSubtitle, actionbarComponent)
            return
        }
        startFirstVisitTasks(player, stop, timings, translatedTitle, translatedSubtitle, actionbarComponent)
    }

    private fun startAlwaysShowTasks(
        player: Player,
        stop: Stop,
        timings: TitleTimings,
        translatedTitle: () -> String,
        translatedSubtitle: () -> String,
        actionbarComponent: () -> Array<BaseComponent>,
    ) {
        val playerId = player.uniqueId

        val actionBarTaskId =
            SchedulerUtil.entityRun(
                plugin,
                player,
                {
                    // 检查任务是否仍然存在于Map中，如果不存在说明已被外部取消
                    // 不在这里取消任务，让外部的 PlayerMoveEvent 来处理
                    if (actionBarTasks.containsKey(playerId) && canDisplayAt(player, stop)) {
                        player.spigot().sendMessage(ChatMessageType.ACTION_BAR, *actionbarComponent())
                    }
                },
                0L,
                ACTION_BAR_INTERVAL_TICKS,
            )
        if (actionBarTaskId != null) {
            actionBarTasks[playerId] = actionBarTaskId
        }

        val titleTaskId =
            SchedulerUtil.entityRun(
                plugin,
                player,
                {
                    if (continuousInfoTasks.containsKey(playerId) && canDisplayAt(player, stop)) {
                        player.sendTitle(
                            translatedTitle(),
                            translatedSubtitle(),
                            timings.continuousFadeIn,
                            timings.continuousStay,
                            timings.continuousFadeOut,
                        )
                    }
                },
                0L,
                timings.interval.toLong(),
            )
        if (titleTaskId != null) {
            continuousInfoTasks[playerId] = titleTaskId
        }
    }

    private fun startFirstVisitTasks(
        player: Player,
        stop: Stop,
        timings: TitleTimings,
        translatedTitle: () -> String,
        translatedSubtitle: () -> String,
        actionbarComponent: () -> Array<BaseComponent>,
    ) {
        val playerId = player.uniqueId
        val metaKey = "metro_first_run_" + stop.id
        if (player.getMetadata(metaKey).isNotEmpty()) {
            return
        }
        player.setMetadata(metaKey, FixedMetadataValue(plugin, true))

        if (isInMetroMinecart(player)) {
            return
        }

        player.sendTitle(
            translatedTitle(),
            translatedSubtitle(),
            timings.singleFadeIn,
            timings.singleStay,
            timings.singleFadeOut,
        )

        val totalDisplayTime = timings.singleStay + timings.singleFadeOut
        val actionBarTaskId =
            SchedulerUtil.entityRun(
                plugin,
                player,
                object : Runnable {
                    private var count = 0
                    private val maxCount = totalDisplayTime / TICKS_PER_SECOND + 1

                    override fun run() {
                        if (!player.isOnline || count >= maxCount || !stop.isInStop(player.location)) {
                            cancelActionBarTask(playerId)
                            return
                        }
                        if (!isInMetroMinecart(player)) {
                            player.spigot().sendMessage(ChatMessageType.ACTION_BAR, *actionbarComponent())
                        }
                        count++
                    }
                },
                0L,
                ACTION_BAR_INTERVAL_TICKS,
            )
        if (actionBarTaskId != null) {
            actionBarTasks[playerId] = actionBarTaskId
        }
    }

    private fun canDisplayAt(player: Player, stop: Stop): Boolean =
        player.isOnline && stop.isInStop(player.location) && !isInMetroMinecart(player)

    private fun buildBoardableRouteSummary(stop: Stop, boardableLines: List<Line>): String {
        val routes = ArrayList<String>()
        for (line in boardableLines) {
            val nextStopId = line.getNextStopId(stop.id)
            val nextStop = if (nextStopId == null) null else plugin.stopManager.getStop(nextStopId)
            val nextStopName = nextStop?.name ?: nextStopId.orEmpty()
            val lineName = line.color + line.name
            var route =
                plugin.languageManager.getMessage(
                    "interact.multi_line_route",
                    mapOf("line_name" to lineName, "next_stop_name" to nextStopName),
                )
            val etaSeconds = estimateEtaSeconds(line, stop)
            if (etaSeconds >= 0) route += " &7(${formatEta(etaSeconds)})"
            routes.add(route)
        }
        return routes.joinToString("&7 | ")
    }

    private fun isInMetroMinecart(player: Player): Boolean {
        val minecart = player.vehicle as? Minecart ?: return false
        val minecartKey = MetroConstants.getMinecartKey()
        if (minecartKey != null && minecart.persistentDataContainer.has(minecartKey, PersistentDataType.BYTE)) {
            return true
        }
        return plugin.lineServiceManager?.getTrainByMinecart(minecart.uniqueId) != null
    }

    private fun translateStopTemplate(
        template: String,
        line: Line,
        stop: Stop,
        lastStop: Stop?,
        nextStop: Stop?,
        terminalStop: Stop?,
        lineManager: org.cubexmc.metro.manager.LineManager,
    ): String =
        ChatColor.translateAlternateColorCodes(
            '&',
            replaceStopTemplate(template, line, stop, lastStop, nextStop, terminalStop, lineManager),
        )

    private fun replaceStopTemplate(
        template: String,
        line: Line,
        stop: Stop,
        lastStop: Stop?,
        nextStop: Stop?,
        terminalStop: Stop?,
        lineManager: org.cubexmc.metro.manager.LineManager,
    ): String {
        val result = TextUtil.replacePlaceholders(template, line, stop, lastStop, nextStop, terminalStop, lineManager)
        val etaSeconds = estimateEtaSeconds(line, stop)
        return result
            .replace("{eta_seconds}", if (etaSeconds >= 0) etaSeconds.toString() else "--")
            .replace("{eta_formatted}", if (etaSeconds >= 0) formatEta(etaSeconds) else "--:--")
    }

    private fun estimateEtaSeconds(line: Line?, stop: Stop?): Int {
        if (line == null || stop == null || !line.isServiceEnabled || plugin.lineServiceManager == null) return -1
        return plugin.lineServiceManager!!.estimateNextEtaSeconds(line.id, stop.id)
    }

    private fun formatEta(seconds: Int): String {
        val safeSeconds = max(0, seconds)
        return String.format("%02d:%02d", safeSeconds / 60, safeSeconds % 60)
    }

    /**
     * 取消显示持续信息的任务
     */
    private fun cancelContinuousInfoTask(playerId: UUID) {
        continuousInfoTasks.remove(playerId)?.let { SchedulerUtil.cancelTask(it) }
    }

    /**
     * 取消ActionBar显示任务
     */
    private fun cancelActionBarTask(playerId: UUID) {
        actionBarTasks.remove(playerId)?.let { SchedulerUtil.cancelTask(it) }
    }

    /**
     * 插件关闭时主动清理所有显示任务和缓存状态
     */
    fun shutdown() {
        for (taskId in continuousInfoTasks.values) {
            SchedulerUtil.cancelTask(taskId)
        }
        for (taskId in actionBarTasks.values) {
            SchedulerUtil.cancelTask(taskId)
        }
        continuousInfoTasks.clear()
        actionBarTasks.clear()
        playerInStopMap.clear()
    }

    /**
     * 标题显示时序。连续显示时禁用淡入淡出避免闪烁，并保证停留时间覆盖刷新间隔。
     */
    private class TitleTimings(
        val interval: Int,
        val alwaysShow: Boolean,
        val continuousFadeIn: Int,
        val continuousStay: Int,
        val continuousFadeOut: Int,
        val singleFadeIn: Int,
        val singleStay: Int,
        val singleFadeOut: Int,
    ) {
        companion object {
            fun from(config: org.cubexmc.metro.config.ConfigFacade): TitleTimings {
                val interval = config.getStopContinuousInterval()
                val alwaysShow = config.isStopContinuousAlways()
                val configuredFadeIn = config.getStopContinuousFadeIn()
                val configuredStay = config.getStopContinuousStay()
                val configuredFadeOut = config.getStopContinuousFadeOut()

                return TitleTimings(
                    interval = interval,
                    alwaysShow = alwaysShow,
                    continuousFadeIn = if (alwaysShow) 0 else configuredFadeIn,
                    continuousStay = if (alwaysShow) max(configuredStay, interval + 1) else configuredStay,
                    continuousFadeOut = if (alwaysShow) 0 else configuredFadeOut,
                    singleFadeIn = configuredFadeIn,
                    singleStay = configuredStay,
                    singleFadeOut = configuredFadeOut,
                )
            }
        }
    }

    private companion object {
        const val ACTION_BAR_INTERVAL_TICKS = 20L
        const val TICKS_PER_SECOND = 20
    }
}
